File: test_contest/solutions/ladders.cpp
// Solution to ladders from 2017 UCF HSPT
// Uses a breath first search to mark every square as visited or not visited
#include <iostream>
#include <string>
#include <queue>

using namespace std;

int main()
{
    int T;
    cin >> T;
    for(int t = 1; t<=T; t++)
    {
        int n, m, d;
        cin >> n >> m >> d;
        // Read in the board and store the starting and ending squares in si/sj/sk and ei/ej/ek.
        char board[d][n][m];
        int si = -1, sj = -1, sk = -1;
        int ei = -1, ej = -1, ek = -1;
        int vis[d][n][m];
        for(int i = 0; i<d; i++)
        {
            for(int j = 0; j<n; j++)
            {
                string s;
                cin >> s;
                for(int k = 0; k<m; k++)
                {
                    vis[i][j][k] = 0;
                    board[i][j][k] = s[k];
                    if(board[i][j][k] == 'S')
                    {
                        si = i; sj = j; sk = k;
                    }
                    else if(board[i][j][k] == 'E')
                    {
                        ei = i; ej = j; ek = k;
                    }

                }
            }
        }
        // When moving around on a level, the options for movements are (0, +1), (0, -1), (+1, 0), and (-1, 0)
        int dj[] = {0, 0, 1, -1};
        int dk[] = {1, -1, 0, 0};
        
        // Perform a breadth-first search of all locations that can be visited.
        // This involves putting each location into a queue, and when it is removed from the queue,
        // figuring out which squares can be reached from there in a single move
        std::queue<int> qi, qj, qk;
        qi.push(si);
        qj.push(sj);
        qk.push(sk);
        vis[si][sj][sk] = 1;
        while(qi.size())
        {
            // Get a location from the queue to process
            int ati = qi.front();
            int atj = qj.front();
            int atk = qk.front();
            qi.pop(); qj.pop(); qk.pop();
            // Case 1: Chute - try to move down to the bottom of the chute
            if(board[ati][atj][atk] == '*')
            {
                int ni = ati, nj = atj, nk = atk;
                while(ni < d - 1 && board[ni+1][atj][nk] == '*') ni++;
                if(ati != ni)
                {
                    if(!vis[ni][nj][nk])
                    {
                        vis[ni][nj][nk] = 1;
                        qi.push(ni);
                        qj.push(nj);
                        qk.push(nk);
                    }
                    continue; // We were at a chute but not the bottom so the only choice was down
                }
            }
            // Case 2: Ladder - try to move up or down by 1
            if(board[ati][atj][atk] == '#')
            {
                int ni = ati - 1, nj = atj, nk = atk;
                if(ni >= 0 && board[ni][nj][nk] == '#' && !vis[ni][nj][nk])
                {
                    vis[ni][nj][nk] = 1;
                    qi.push(ni);
                    qj.push(nj);
                    qk.push(nk);
                }
                ni = ati + 1; nj = atj; nk = atk;
                if(ni < d && board[ni][nj][nk] == '#' && !vis[ni][nj][nk])
                {
                    vis[ni][nj][nk] = 1;
                    qi.push(ni);
                    qj.push(nj);
                    qk.push(nk);
                }
            }
            // Case 3: Move around on the same level
            for(int move = 0; move < 4; move++)
            {
                int nj = atj + dj[move], nk = atk + dk[move], ni = ati;
                if(nk >= 0 && nk < m && nj >= 0 && nj < n && !vis[ni][nj][nk])
                {
                    vis[ni][nj][nk] = 1;
                    qi.push(ni);
                    qj.push(nj);
                    qk.push(nk);
                }
            }
            
        }
        string ans = "No";
        if(vis[ei][ej][ek]) ans = "Yes";
        cout << "Map #" << t << ": " << ans << endl;
    }
    return 0;
}

File: test_contest/solutions/ladders.py
# Written by Will Cromar
# Python 3.6

from collections import deque

# DX/DY array. In order of precedence, we'll move
# left, right, up, and down
DX = [-1, 1, 0, 0]
DY = [0, 0, -1, 1]

# Constants for types of spaces
EMPTY = '.'
LADDER = '#'
CHUTE = '*'
START = 'S'
EXIT = 'E'


# Primary business functions. Includes several helpers
def solve(grid, l, w, h):
    # Generator that gives all adjacent spots to the given one
    def adjacent(x, y, z):
        # print("call to adj w/", (x, y, z))
        for dx, dy in zip(DX, DY):
            # print(x + dx, y + dy, z)
            yield x + dx, y + dy, z

    # True if the result is in bounds, false otherwise
    def inBounds(x, y, z):
        return x >= 0 and y >= 0 and z >= 0 and x < w and y < l and z < h

    # Gives all adjacent nodes that are in bounds
    def moves(x, y, z):
        # Get all adjacent nodes and then filter them
        return filter(lambda triplet: inBounds(*triplet), adjacent(x, y, z))

    # Follow a chute to the bottom
    def chuteBottom(x, y, z):
        # While the spot below us is still a chute, keep moving down levels
        while z > 0 and grid[z - 1][y][x] == CHUTE:
            z -= 1

        return x, y, z

    # Gives vertically adjacent spots that have ladders
    def ladderSpots(x, y, z):
        # If there's a ladder above us, yield it
        if z < h - 1 and grid[z + 1][y][x] == LADDER:
            yield x, y, z + 1

        # Likewise, yield ladders below us
        if z > 0 and grid[z - 1][y][x] == LADDER:
            yield x, y, z - 1

    # Searches for the start symbol and returns an ordered triplet
    # or None
    def findStart():
        for z, level in enumerate(grid):
            for y, row in enumerate(level):
                if START in row:
                    return row.find(START), y, z

    # Finally, we can move on to our breadth-first search
    # This will store which spots we've seen before
    seen = set()

    # Will store an ordered triplet and a number of steps
    q = deque()

    # Find the starting position and add it to the queue w/ 0 steps
    start = findStart()
    seen.add(start)
    q.append(start)

    # Run BFS until the queue empties
    while len(q) > 0:
        # Unpack the data from the queue
        triplet = q.popleft()
        x, y, z = triplet

        # If we've found the exit, we're done!
        if grid[z][y][x] == EXIT:
            return True

        # If we stepped on a chute, fall
        if grid[z][y][x] == CHUTE:
            next = chuteBottom(x, y, z)
            if next not in seen:
                seen.add(next)
                q.appendleft(next)

            # If we're not at the bottom of the chute, then we
            # can't move anywhere else
            if next != triplet:
                continue

        # If we stepped on a ladder, see where we can go
        if grid[z][y][x] == LADDER:
            for next in ladderSpots(x, y, z):
                if next not in seen:
                    seen.add(next)
                    q.append(next)

        # Otherwise, explore our "slice" of the map
        for next in moves(x, y, z):
            if next not in seen:
                seen.add(next)
                q.append(next)

    return False

# Number of test cases
n = int(input())

# Process each test case
for case in range(n):
    # Get the dimensions of the  grid
    l, w, h = map(int, input().strip().split())

    # Read it in
    grid = []
    for _ in range(h):
        slice = []
        for _ in range(l):
            slice.append(input().strip())

        grid.append(slice)

    # I reverse the list to make z = 0 correspond to the lowest
    # slice on the map
    grid.reverse()

    # Determine if there is a path from start to end
    ans = solve(grid, l, w, h)
    # If there is, print "Yes", otherwise, "No"
    print("Map #%d: %s" % (case + 1, "Yes" if ans else "No"))
